Pulse meter internal filter mode

cstaahl opened the pull request:
# What does this implement/fix? 

Adds an optional configuration variable "internal_filter_mode" to the "pulse_meter" sensor. It improves detection of long pulses as requested by [Pulse Meter debounce rising signal #1354](https://github.com/esphome/feature-requests/issues/1354). The default value of this configuration variable, "EDGE", retains identical behaviour as before. Setting it to "PULSE" changes the behaviour of the filter logic.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1832

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:
<!--
sensor:
  - platform: pulse_meter
    pin: 14
    unit_of_measurement: 'W'
    name: 'Current power'
    id: my_power
    device_class: "power"
    internal_filter_mode: "PULSE"
    internal_filter: 100ms
    accuracy_decimals: 0
    filters:
      - multiply: 80.
    total:
-->

```yaml

esphome:
  name: energymeter

esp8266:
  board: nodemcuv2

# Enable logging
logger:

# Enable Home Assistant API
api:
  password: ""

ota:
  password: ""

wifi:
  ssid: "!secret wifi_ssid"
  password: "!secret wifi_password"

# Example configuration entry
sensor:
  - platform: pulse_meter
    pin: 14
    unit_of_measurement: 'W'
    name: 'Current power'
    device_class: "power"
    internal_filter_mode: "PULSE"
    internal_filter: 100ms
    accuracy_decimals: 0
    filters:
      - multiply: 80.
    total:
      name: "Electricity Meter"
      unit_of_measurement: "kWh"
      device_class: "energy"
      accuracy_decimals: 2
      filters:
        - multiply: 0.00133333    

# Enable time component to reset energy at midnight
time:
  - platform: sntp
    id: my_time
captive_portal:
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

cstaahl committed:
Moved filter and pulse count logic to loop() from isr

cstaahl committed:
Added "internal_filter_mode" configuration variable

cstaahl committed:
Added PULSE filter_mode functionality

cstaahl committed:
Updated codeowners

cstaahl commented:
Docs update in https://github.com/esphome/esphome-docs/pull/1829

stevebaxter commented on esphome/components/pulse_meter/pulse_meter_sensor.h:
```diff
@@ -8,10 +8,16 @@
 namespace esphome {
 namespace pulse_meter {
 
+enum PulseMeterInternalFilterMode {
```
Maybe consider moving this inside the class and calling it `InternalFilterMode` so you have `PulseMeterSensor::FILTER_EDGE` and `PulseMeterSensor::FILTER_PULSE`.

cstaahl replied:
You are right, that is a more elegant approach. I'll push another commit implementing what you suggested shortly.

stevebaxter reviewed (approved):
Neat! Moving stuff out of the interrupt handler is good as well, that might help solve some other issues people have reported.

One suggestion, but not essential.

pawlizio commented on esphome/components/pulse_meter/pulse_meter_sensor.cpp:
```diff
@@ -18,6 +18,31 @@ void PulseMeterSensor::setup() {
+    if ((now - this->last_detected_edge_us_) >= this->filter_us_) {
+      // Only consider high pulses and "new" edges
+      if (this->isr_pin_.digital_read() && (this->last_detected_edge_us_ > this->last_valid_edge_us_)) {
+        // Don't measure the first valid pulse (we need at least two pulses to measure the width)
```
I tested your code on a gas meter and I had the case, that my heating switches off at 21pm, directly during the tests. However this case an issue occured where the total-counter was continuesly increasing while the pulse/min sensor reports 0.

I think the issue is caused here. In line [51](https://github.com/esphome/esphome/blob/0f696301673323792ccdb8d79ac1f13d2b588087/esphome/components/pulse_meter/pulse_meter_sensor.cpp#L51) `last_valid_edge_us_` will be set 0. In this case your if conditions in the loop will be true, if the sensor stops at HIGH level, so the sensor increments in each loop cycle.

pawlizio replied:
I created a [PR ](https://github.com/cstaahl/esphome/pull/1) with my proposal. I removed line 51, so that `last_valid_edge_us_` will never set to 0. For me this has solved the issue.

cstaahl replied:
Great catch, thanks for the quick fix!

pawlizio committed:
Fix endless loop for total_counter

This is my proposal for the issue in your PR which I mentioned here: https://github.com/esphome/esphome/pull/3082/files#r789984784

cstaahl committed:
Merge pull request #1 from pawlizio/patch-1

Fix endless loop for total_counter

cstaahl committed:
Moved InternalFilterMode enumeration inside the PulseMeterSensor class

pawlizio committed:
Fix micros() overflow

I assume a this should solve that issue with micros() overflowing.

cstaahl committed:
Merge pull request #2 from pawlizio/patch-2

Fix micros() overflow

pawlizio commented:
I still figure out sometimes two increments in a very short time which should not occur. I think it may happen that there is sometimes a noisy signal, so short interrups during HIGH state of a sensor, as shown below:
![image](https://user-images.githubusercontent.com/15065886/150681532-39c2dbca-ecd8-4add-a34c-a54e73db12e4.png)

pawlizio committed:
Add sensor state (#3)

* Add sensor state in order to make behavior symmetric with regard to high or low pulses

cstaahl committed:
Format changes

jesserockz committed:
Merge branch 'dev' into pulse_meter-filter_mode_pulse

OttoWinter committed:
Fix allow enums in classes

OttoWinter reviewed (approved):
LGTM, thanks!

(I don't know the pulse meter internals well, but as the original codeowner approved, this should be fine)

Could you also update the docs (basically just add a line documenting this option), thanks! https://github.com/esphome/esphome-docs/

cstaahl commented:
> LGTM, thanks!
> 
> (I don't know the pulse meter internals well, but as the original codeowner approved, this should be fine)
> 
> Could you also update the docs (basically just add a line documenting this option), thanks! https://github.com/esphome/esphome-docs/

I added a [PR](https://github.com/esphome/esphome-docs/pull/1832) in the docs already. Perhaps I didn't link the two correctly.

jesserockz commented:
> I added a [PR](https://github.com/esphome/esphome-docs/pull/1832) in the docs already. Perhaps I didn't link the two correctly.

It needs to be linked in this PR description :smile:  (I have added above)
